File: app/src/main/java/com/example/energo_monitoring/presenter/utilities/LoadImageManager.java
package com.example.energo_monitoring.presenter.utilities;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.provider.Settings;

import androidx.activity.result.ActivityResultLauncher;
import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

public class LoadImageManager {

    private Context context;
    private ActivityResultLauncher<Intent> takePhotoResult;
    private ActivityResultLauncher<Intent> getPhotoFromGalleryResult;
    private ActivityResultLauncher<Intent> takePermissions;

    public LoadImageManager(Context context, ActivityResultLauncher<Intent> takePhotoResult,
                            ActivityResultLauncher<Intent> getPhotoFromGalleryResult,
                            ActivityResultLauncher<Intent> takePermissions) {
        this.context = context;
        this.takePhotoResult = takePhotoResult;
        this.getPhotoFromGalleryResult = getPhotoFromGalleryResult;
        this.takePermissions = takePermissions;
    }

    public static void takePhoto(ActivityResultLauncher<Intent> launcher){
        Intent cameraIntent = new Intent(android.provider.MediaStore.ACTION_IMAGE_CAPTURE);
        launcher.launch(cameraIntent);
    }

    public void pickFromGallery(){
        if(isPermissionGranted()) {
            Intent galleryIntent = new Intent(Intent.ACTION_GET_CONTENT);
            galleryIntent.setType("image/*");
            galleryIntent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);

            getPhotoFromGalleryResult.launch(galleryIntent);
        } else {
            takePermission();
        }
    }

    public boolean isPermissionGranted(){
        if (Build.VERSION.SDK_INT == Build.VERSION_CODES.R){
            return Environment.isExternalStorageManager();
        } else {
            int readExternalStorage = ContextCompat.checkSelfPermission(context,
                    Manifest.permission.READ_EXTERNAL_STORAGE);
            return readExternalStorage == PackageManager.PERMISSION_GRANTED;
        }
    }

    public void takePermission(){
        if(Build.VERSION.SDK_INT == Build.VERSION_CODES.R){
            try{
                Intent intent = new Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION);
                Uri uri = Uri.fromParts("package", context.getPackageName(), null);
                intent.setData(uri);
                takePermissions.launch(intent);
            } catch (Exception e){
                Intent intent = new Intent();
                intent.setAction(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION);
                takePermissions.launch(intent);
            }
        } else {
            ActivityCompat.requestPermissions((Activity) context,
                    new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, 101);
        }
    }

    public void onRequestPermissionResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults){
        if(grantResults.length > 0){
            if(requestCode == 101){
                boolean readExternalStorage = grantResults[0] == PackageManager.PERMISSION_GRANTED;
                if(readExternalStorage){
                    Intent galleryIntent = new Intent(Intent.ACTION_GET_CONTENT);
                    galleryIntent.setType("image/*");
                    galleryIntent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);

                    getPhotoFromGalleryResult.launch(galleryIntent);
                }
            }
        }
    }
}

File: app/src/main/java/com/example/energy_monitoring/checks/ui/utils/LoadImageManager.kt
package com.example.energy_monitoring.checks.ui.utils

import android.Manifest
import androidx.activity.result.ActivityResultLauncher
import android.content.Intent
import android.os.Build
import android.os.Environment
import androidx.core.content.ContextCompat
import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import android.app.Activity
import android.content.Context
import android.provider.MediaStore
import android.widget.Toast
import androidx.core.content.FileProvider
import android.graphics.Bitmap
import android.net.Uri
import android.provider.Settings
import android.util.Base64
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.lang.Exception
import java.lang.StringBuilder

class LoadImageManager(
    private val context: Context,
    private val getPhotoFromGalleryResult: ActivityResultLauncher<Intent>,
    private val takePermissions: ActivityResultLauncher<Intent>
) {
    private val isPermissionGranted: Boolean
        get() = if (Build.VERSION.SDK_INT == Build.VERSION_CODES.R) {
            Environment.isExternalStorageManager()
        } else {
            val readExternalStorage = ContextCompat.checkSelfPermission(
                context,
                Manifest.permission.READ_EXTERNAL_STORAGE
            )
            readExternalStorage == PackageManager.PERMISSION_GRANTED
        }

    fun pickFromGallery() {
        if (isPermissionGranted) {
            val galleryIntent = Intent(Intent.ACTION_GET_CONTENT)
            galleryIntent.type = "image/*"
            galleryIntent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
            getPhotoFromGalleryResult.launch(galleryIntent)
        } else {
            takePermission()
        }
    }

    fun takePermission() {
        if (Build.VERSION.SDK_INT == Build.VERSION_CODES.R) {
            try {
                val intent = Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION)
                val uri = Uri.fromParts("package", context.packageName, null)
                intent.data = uri
                takePermissions.launch(intent)
            } catch (e: Exception) {
                val intent = Intent()
                intent.action = Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION
                takePermissions.launch(intent)
            }
        } else {
            ActivityCompat.requestPermissions(
                (context as Activity),
                arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE),
                101
            )
        }
    }

    fun onRequestPermissionResult(
        requestCode: Int,
        permissions: Array<String?>,
        grantResults: IntArray
    ) {
        if (grantResults.isNotEmpty()) {
            if (requestCode == 101) {
                val readExternalStorage = grantResults[0] == PackageManager.PERMISSION_GRANTED
                if (readExternalStorage) {
                    val galleryIntent = Intent(Intent.ACTION_GET_CONTENT)
                    galleryIntent.type = "image/*"
                    galleryIntent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
                    getPhotoFromGalleryResult.launch(galleryIntent)
                }
            }
        }
    }

    companion object {
        @JvmStatic
        fun takePhoto(launcher: ActivityResultLauncher<Intent>, path: Uri?) {
            val cameraIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
            cameraIntent.putExtra(MediaStore.EXTRA_OUTPUT, path)
            launcher.launch(cameraIntent)
        }

        @JvmStatic
        fun getPhotoUri(context: Context, dataId: Int, saveRealPath: (String) -> Unit): Uri {
            var file: File? = null
            try {
                val folder = Utils.getDataIdFolder(context, dataId)
                file = File.createTempFile("photo", ".png", folder)
                saveRealPath(file.canonicalPath)
            } catch (e: IOException) {
                e.printStackTrace()
                Toast.makeText(context, e.toString(), Toast.LENGTH_LONG).show()
            }
            return FileProvider.getUriForFile(
                context,
                context.applicationContext.packageName + ".provider", file!!
            )
        }

        fun getBase64FromPath(context: Context, paths: String): String {
            val result = StringBuilder()
            for (path in paths.split(";").toTypedArray()) {
                if (path.isEmpty()) continue
                var bitmap: Bitmap? = null
                try {
                    bitmap = MediaStore.Images.Media.getBitmap(
                        context.contentResolver,
                        Uri.parse(path)
                    )
                } catch (e: IOException) {
                    e.printStackTrace()
                }
                val byteArrayOutputStream = ByteArrayOutputStream()
                if (bitmap == null) {
                    Toast.makeText(
                        context,
                        "Произошла ошибка! Bitmap равен null!",
                        Toast.LENGTH_LONG
                    ).show()
                    return ""
                }
                bitmap.compress(Bitmap.CompressFormat.JPEG, 100, byteArrayOutputStream)
                val byteArray = byteArrayOutputStream.toByteArray()
                val encoded = Base64.encodeToString(byteArray, Base64.DEFAULT)
                result.append(encoded).append(";")
            }
            if (result.isNotEmpty()) result.substring(0, result.length - 1)
            return result.toString()
        }
    }
}